hibernate-core/src/main/java/org/hibernate/query/sqm/sql/internal/StandardSqmInsertTranslator.java

```diff
@@ -0,0 +1,233 @@
+ * See the lgpl.txt file in the root directory or http://www.gnu.org/licenses/lgpl-2.1.html
+ */
+package org.hibernate.query.sqm.sql.internal;
+
+import org.hibernate.HibernateException;
+import org.hibernate.LockMode;
+import org.hibernate.internal.util.collections.CollectionHelper;
+import org.hibernate.persister.entity.EntityPersister;
+import org.hibernate.query.NavigablePath;
+import org.hibernate.query.spi.QueryOptions;
+import org.hibernate.query.spi.QueryParameterBindings;
+import org.hibernate.query.sqm.internal.DomainParameterXref;
+import org.hibernate.query.sqm.sql.BaseSqmToSqlAstConverter;
+import org.hibernate.query.sqm.sql.SqmInsertTranslation;
+import org.hibernate.query.sqm.sql.SqmInsertTranslator;
+import org.hibernate.query.sqm.tree.cte.SqmCteStatement;
+import org.hibernate.query.sqm.tree.domain.SqmPath;
+import org.hibernate.query.sqm.tree.expression.SqmExpression;
+import org.hibernate.query.sqm.tree.insert.SqmInsertSelectStatement;
+import org.hibernate.query.sqm.tree.insert.SqmInsertStatement;
+import org.hibernate.query.sqm.tree.insert.SqmInsertValuesStatement;
+import org.hibernate.query.sqm.tree.insert.SqmValues;
+import org.hibernate.query.sqm.tree.select.SqmSelectStatement;
+import org.hibernate.query.sqm.tree.select.SqmSelection;
+import org.hibernate.sql.ast.spi.SqlAstCreationContext;
+import org.hibernate.sql.ast.spi.SqlAstCreationState;
+import org.hibernate.sql.ast.tree.cte.CteStatement;
+import org.hibernate.sql.ast.tree.expression.Expression;
+import org.hibernate.sql.ast.tree.from.TableGroup;
+import org.hibernate.sql.ast.tree.insert.InsertStatement;
+import org.hibernate.sql.ast.tree.insert.Values;
+import org.hibernate.sql.ast.tree.select.QuerySpec;
+import org.hibernate.sql.ast.tree.select.SelectStatement;
+import org.hibernate.sql.ast.tree.update.Assignable;
+import org.hibernate.sql.results.graph.DomainResult;
+import org.hibernate.sql.results.graph.DomainResultCreationState;
+
+import java.util.List;
+
+/**
+ * @author Steve Ebersole
+ */
+public class StandardSqmInsertTranslator
+		extends BaseSqmToSqlAstConverter
+		implements SqmInsertTranslator, DomainResultCreationState {
+
+	private final List<DomainResult> domainResults = CollectionHelper.arrayList( 10 );
+
+	public StandardSqmInsertTranslator(
+			SqlAstCreationContext creationContext,
+			QueryOptions queryOptions,
+			DomainParameterXref domainParameterXref,
+			QueryParameterBindings domainParameterBindings) {
+		super( creationContext, queryOptions, domainParameterXref, domainParameterBindings );
+	}
+
+	@Override
+	public SqmInsertTranslation translate(SqmInsertStatement sqmStatement) {
+		InsertStatement sqlAst;
+		if ( sqmStatement instanceof SqmInsertSelectStatement ) {
+			sqlAst = visitInsertSelectStatement( (SqmInsertSelectStatement) sqmStatement );
+		}
+		else {
+			sqlAst = visitInsertValuesStatement( (SqmInsertValuesStatement) sqmStatement );
+		}
+		return new SqmInsertTranslation( sqlAst, getJdbcParamsBySqmParam() );
```

Comment:
@sebersole if you absolutely insist, we could have separate `StandardSqmInsertValuesTranslator` and `StandardSqmInsertSelectTranslator`, but it just seems all a bit like overkill.

To be honest I even really don't see the need for both `SqmInsertSelectStatement` and `SqmInsertValuesStatement`. I would make them one class.

Reply:
I agree a single translator is enough.

We could consolidate the 2 statement types but that would mean having a commonality between SqmQuerySpec and whatever you use to model the values constructor.  So I'm not against this part, though its not a big deal to me